addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/actions/ProcessRetentionEventAction.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * @since 11.1
+ */
+public class ProcessRetentionEventAction implements StreamProcessorTopology {
+
+    public static final String ACTION_NAME = "processRetentionEvent";
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+        return Topology.builder()
+                       .addComputation(ProcessRetentionEventComputation::new,
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class ProcessRetentionEventComputation extends AbstractBulkComputation {
+
+        private static final Logger log = LogManager.getLogger(ProcessRetentionEventComputation.class);
+
+        protected boolean disableAudit;
+
+        protected RetentionManager retentionManager;
+
+        public ProcessRetentionEventComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        public void startBucket(String bucketKey) {
+            BulkCommand command = getCurrentCommand();
+            Serializable auditParam = command.getParam(NXAuditEventsService.DISABLE_AUDIT_LOGGER);
+            disableAudit = auditParam != null && Boolean.parseBoolean(auditParam.toString());
+            retentionManager = Framework.getService(RetentionManager.class);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+            for (DocumentModel ruleDoc : loadDocuments(session, ids)) {
+                if (!ruleDoc.hasFacet(RetentionConstants.RETENTION_RULE_FACET)) {
+                    log.debug("Document {} is not a retention rule, ignoring ...", ruleDoc::getPathAsString);
+                    continue;
+                }
+                RetentionRule rule = ruleDoc.getAdapter(RetentionRule.class);
+                scheduleInputEventBasedRule(rule);
+            }
+        }
+
+        protected void scheduleInputEventBasedRule(RetentionRule rule) {
+            if (!rule.isEnabled() || !rule.isEventBased()) {
+                throw new IllegalArgumentException("Rule is disabled or not event-based");
+            }
+            BulkService bulkService = Framework.getService(BulkService.class);
+            RepositoryService repositoryService = Framework.getService(RepositoryService.class);
+            StringBuilder query = new StringBuilder(RetentionConstants.RULE_RECORD_DOCUMENT_QUERY);
+            query.append(" AND ") //
+                 .append(RetentionConstants.RECORD_RULE_IDS_PROP) //
+                 .append(" = '" + rule.getDocument().getId() + "'");
+            for (String repositoryName : repositoryService.getRepositoryNames()) {
+                BulkCommand command = new BulkCommand.Builder(EvalInputEventBasedRuleAction.ACTION_NAME,
+                        query.toString()).user(SecurityConstants.SYSTEM_USERNAME).repository(repositoryName).build();
+                bulkService.submit(command);
```

Comment:
`org.nuxeo.ecm.core.bulk.message.BulkCommand.Builder#Builder(java.lang.String, java.lang.String)` is a deprecated constructor, we can replace it by: 

```
BulkCommand command = new BulkCommand.Builder(EvalInputEventBasedRuleAction.ACTION_NAME,
                        query.toString(), SecurityConstants.SYSTEM_USERNAME).repository(repositoryName).build();
```